In [ ]:
import pandas as pd
import numpy as np
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeClassifier
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier


ingest = pd.read_csv('sample_data/amazon.csv')



In [ ]:
ingest = ingest.dropna()

In [ ]:
ingest.head()


,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count,about_product,user_id,user_name,review_id,review_title,review_content,img_link,product_link
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories|Accessories&Peripherals|...,₹399,"₹1,099",64%,4.2,"24,269",High Compatibility : Compatible With iPhone 12...,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories|Accessories&Peripherals|...,₹199,₹349,43%,4.0,"43,994","Compatible with all Type C enabled devices, be...","AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...","ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...","RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...","A Good Braided Cable for Your Type C Device,Go...",I ordered this cable to connect my phone to An...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Ambrane-Unbreakable-Char...
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories|Accessories&Peripherals|...,₹199,"₹1,899",90%,3.9,"7,928",【 Fast Charger& Data Sync】-With built-in safet...,"AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...","Kunal,Himanshu,viswanath,sai niharka,saqib mal...","R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...","Good speed for earlier versions,Good Product,W...","Not quite durable and sturdy,https://m.media-a...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Sounce-iPhone-Charging-C...
3,B08HDJ86NZ,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,Computers&Accessories|Accessories&Peripherals|...,₹329,₹699,53%,4.2,"94,363",The boAt Deuce USB 300 2 in 1 cable is compati...,"AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...","Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...","R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...","Good product,Good one,Nice,Really nice product...","Good product,long wire,Charges good,Nice,I bou...",https://m.media-amazon.com/images/I/41V5FtEWPk...,https://www.amazon.in/Deuce-300-Resistant-Tang...
4,B08CF3B7N1,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,Computers&Accessories|Accessories&Peripherals|...,₹154,₹399,61%,4.2,"16,905",[CHARGE & SYNC FUNCTION]- This cable comes wit...,"AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...","rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...","R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...","As good as original,Decent,Good one for second...","Bought this instead of original apple, does th...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Portronics-Konnect-POR-1...


In [ ]:
ingest['discounted_price']= ingest['discounted_price'].str.replace('₹','').str.replace(',','').astype(float)
ingest['actual_price']= ingest['actual_price'].str.replace('₹','').str.replace(',','').astype(float)
ingest['discount_percentage'] = ingest['discount_percentage'].str.replace('%','').astype(float)
ingest['rating_count'] = ingest['rating_count'].str.replace(',','').astype(float)

ingest.head()

,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count,about_product,user_id,user_name,review_id,review_title,review_content,img_link,product_link
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories|Accessories&Peripherals|...,399.0,1099.0,64.0,4.2,24269.0,High Compatibility : Compatible With iPhone 12...,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories|Accessories&Peripherals|...,199.0,349.0,43.0,4.0,43994.0,"Compatible with all Type C enabled devices, be...","AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...","ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...","RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...","A Good Braided Cable for Your Type C Device,Go...",I ordered this cable to connect my phone to An...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Ambrane-Unbreakable-Char...
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories|Accessories&Peripherals|...,199.0,1899.0,90.0,3.9,7928.0,【 Fast Charger& Data Sync】-With built-in safet...,"AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...","Kunal,Himanshu,viswanath,sai niharka,saqib mal...","R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...","Good speed for earlier versions,Good Product,W...","Not quite durable and sturdy,https://m.media-a...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Sounce-iPhone-Charging-C...
3,B08HDJ86NZ,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,Computers&Accessories|Accessories&Peripherals|...,329.0,699.0,53.0,4.2,94363.0,The boAt Deuce USB 300 2 in 1 cable is compati...,"AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...","Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...","R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...","Good product,Good one,Nice,Really nice product...","Good product,long wire,Charges good,Nice,I bou...",https://m.media-amazon.com/images/I/41V5FtEWPk...,https://www.amazon.in/Deuce-300-Resistant-Tang...
4,B08CF3B7N1,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,Computers&Accessories|Accessories&Peripherals|...,154.0,399.0,61.0,4.2,16905.0,[CHARGE & SYNC FUNCTION]- This cable comes wit...,"AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...","rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...","R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...","As good as original,Decent,Good one for second...","Bought this instead of original apple, does th...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Portronics-Konnect-POR-1...


In [ ]:
ingest['rating'] = pd.to_numeric(ingest['rating'], errors='coerce')

ingest['label']= ingest['rating'].apply(lambda x: 1 if x>=4 else 0)

ingest['label'].value_counts()

,count
label,
1,1109
0,354


In [ ]:
X = ingest[['discounted_price', 'actual_price','rating_count' ]]
Y = ingest['label']

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y,test_size=0.2,stratify=Y,random_state=42)


In [ ]:
X.isnull().sum()

,0
discounted_price,0
actual_price,0
rating_count,0


In [ ]:
len(ingest
    ) #ok so there are a total of 1465 rows ...there is 2 rows with Nan value..so i am dropping them..coz it will only be 0.13% of missing data

1463

In [ ]:
model1 = LogisticRegression(max_iter=500)
model1.fit(X_train, Y_train)

LogisticRegression(max_iter=500)

In [ ]:
Y_testpred = model1.predict(X_test)

print('Test Accuracy:', accuracy_score(Y_test, Y_testpred) * 100, '%')
print('Confusion Matrix:\n', confusion_matrix(Y_test, Y_testpred))

Test Accuracy: 75.76791808873719 %
Confusion Matrix:
 [[  0  71]
 [  0 222]]


In [ ]:
#75% is kinda okish but not so satisfactory

In [ ]:
#yeah so i used logistic regression for it but for the 2nd model might check with knn decision tree

#well the thing is that i cant or maybe i still dont know how to access to notebooks together on colab yet so i am testing everything that will be done
#on notebook2 in below cells i will implement those algorithms in the main model that will show highest accuracy


In [ ]:
df = ingest.copy()


In [ ]:
df['price_diff'] = df['actual_price']-df['discounted_price']
df['discounted_ratio'] = df['discounted_price']/df['actual_price']
df['log_rating_count'] = np.log(df['rating_count'])
df.head()

,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count,about_product,user_id,user_name,review_id,review_title,review_content,img_link,product_link,label,price_diff,discounted_ratio,log_rating_count
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories|Accessories&Peripherals|...,399.0,1099.0,64.0,4.2,24269.0,High Compatibility : Compatible With iPhone 12...,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...,1,700.0,0.363057,10.096955
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories|Accessories&Peripherals|...,199.0,349.0,43.0,4.0,43994.0,"Compatible with all Type C enabled devices, be...","AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...","ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...","RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...","A Good Braided Cable for Your Type C Device,Go...",I ordered this cable to connect my phone to An...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Ambrane-Unbreakable-Char...,1,150.0,0.570201,10.691809
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories|Accessories&Peripherals|...,199.0,1899.0,90.0,3.9,7928.0,【 Fast Charger& Data Sync】-With built-in safet...,"AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...","Kunal,Himanshu,viswanath,sai niharka,saqib mal...","R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...","Good speed for earlier versions,Good Product,W...","Not quite durable and sturdy,https://m.media-a...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Sounce-iPhone-Charging-C...,0,1700.0,0.104792,8.978156
3,B08HDJ86NZ,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,Computers&Accessories|Accessories&Peripherals|...,329.0,699.0,53.0,4.2,94363.0,The boAt Deuce USB 300 2 in 1 cable is compati...,"AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...","Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...","R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...","Good product,Good one,Nice,Really nice product...","Good product,long wire,Charges good,Nice,I bou...",https://m.media-amazon.com/images/I/41V5FtEWPk...,https://www.amazon.in/Deuce-300-Resistant-Tang...,1,370.0,0.470672,11.454904
4,B08CF3B7N1,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,Computers&Accessories|Accessories&Peripherals|...,154.0,399.0,61.0,4.2,16905.0,[CHARGE & SYNC FUNCTION]- This cable comes wit...,"AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...","rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...","R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...","As good as original,Decent,Good one for second...","Bought this instead of original apple, does th...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Portronics-Konnect-POR-1...,1,245.0,0.385965,9.735365


In [ ]:
X2 = df[['discounted_price','actual_price','price_diff','discounted_ratio','log_rating_count']]
Y2 = df['label']

In [ ]:
X2 = df[['discounted_price','actual_price','price_diff','discounted_ratio','log_rating_count']]
Y2 = df['label']

X_train2, X_test2, Y_train2, Y_test2 = train_test_split(X2, Y2,test_size=0.2,stratify=Y2,random_state=42)

len(X_train2), len(X_test2)


(1170, 293)

In [ ]:

len(Y_train2), len(Y_test2)

(1170, 293)

In [ ]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
print(type(scaler))

<class 'sklearn.preprocessing._data.StandardScaler'>


In [ ]:
scaler = StandardScaler()
X_train_Scaled = scaler.fit_transform(X_train2)
X_test_Scaled = scaler.transform(X_test2)

print(X_train_Scaled[:5])
print(X_train2[:5])

[[-0.36124721 -0.41119154 -0.42925454  0.68257562  1.32524439]
 [-0.23657863 -0.09347214  0.1365086  -0.87753862  1.72331471]
 [-0.4235815  -0.40191183 -0.31274302 -1.73732984  0.87827992]
 [-0.26428275 -0.22603934 -0.13640126 -0.40964885  1.86143465]
 [-0.39587737 -0.43735147 -0.43891135  0.27961282 -1.48660792]]
     discounted_price  actual_price  price_diff  discounted_ratio  \
589             599.0         895.0       296.0          0.669274   
585            1499.0        4490.0      2991.0          0.333853   
642             149.0        1000.0       851.0          0.149000   
591            1299.0        2990.0      1691.0          0.434448   
71              349.0         599.0       250.0          0.582638   

     log_rating_count  
589         11.023763  
585         11.827400  
642         10.121418  
591         12.106241  
71           5.347108  


KNN

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

knn_nigga = KNeighborsClassifier(n_neighbors=5)
knn_nigga.fit(X_train_Scaled, Y_train2)

KNeighborsClassifier()

In [ ]:
Y_pred_knn = knn_nigga.predict(X_test_Scaled)
print("KNN Test Accuracy:", accuracy_score(Y_test2, Y_pred_knn) * 100, '%')

KNN Test Accuracy: 76.10921501706484 %


Decision Tree

In [ ]:
dt_nigga = DecisionTreeClassifier(max_depth=2, random_state=42)
dt_nigga.fit(X_train2, Y_train2)

y_pred_dt = dt_nigga.predict(X_test2)
print("DT Test Accuracy:", accuracy_score(Y_test2, y_pred_dt) * 100, '%')

#i cant understand first knn it gave accuracy of 73% now i used decision tree still i got just 76% of accuracy
#i will try to balance overlifting and try using hyper parameters

DT Test Accuracy: 75.76791808873719 %


In [ ]:
for depth in [3,4, 5,6, 7, 8,9, 10, None]:
    model = DecisionTreeClassifier(max_depth=depth,min_samples_leaf=2,min_samples_split=5,random_state=42)
    model.fit(X_train2, Y_train2)
    acc = model.score(X_test2, Y_test2)
    print(f"Depth {depth}: {acc}")

Depth 3: 0.7474402730375427
Depth 4: 0.764505119453925
Depth 5: 0.7474402730375427
Depth 6: 0.7508532423208191
Depth 7: 0.764505119453925
Depth 8: 0.7610921501706485
Depth 9: 0.7406143344709898
Depth 10: 0.757679180887372
Depth None: 0.7167235494880546


i checked with the max_depth # 7 is giving highest accuracy
now checking for min_sampls_leaf..although the lower the no the better it gives accuracy..but lets try

highest accuracy recorded: 76.45

In [ ]:
for leaf in [1, 2, 3, 4, 5, 10]:
    model = DecisionTreeClassifier(max_depth=8,min_samples_leaf=leaf,min_samples_split=5,random_state=42)
    model.fit(X_train2, Y_train2)
    acc = model.score(X_test2, Y_test2)
    print(f"Leaf {leaf}: {acc}")

Leaf 1: 0.7679180887372014
Leaf 2: 0.7610921501706485
Leaf 3: 0.7610921501706485
Leaf 4: 0.7747440273037542
Leaf 5: 0.7747440273037542
Leaf 10: 0.764505119453925


till now depth: 7, min_samples_leaf: 4 giving high accuracy

highest accuracy: 77.47%

In [ ]:
for split in [2, 3, 4, 5, 6, 7, 8, 9, 10]:
    model = DecisionTreeClassifier(max_depth=8,min_samples_leaf=4,min_samples_split=split,random_state=42)
    model.fit(X_train2, Y_train2)
    acc = model.score(X_test2, Y_test2)
    print(f"Split {split}: {acc}")

Split 2: 0.7747440273037542
Split 3: 0.7747440273037542
Split 4: 0.7747440273037542
Split 5: 0.7747440273037542
Split 6: 0.7747440273037542
Split 7: 0.7747440273037542
Split 8: 0.7747440273037542
Split 9: 0.7747440273037542
Split 10: 0.7781569965870307


In [ ]:
model = DecisionTreeClassifier(max_depth=7,min_samples_leaf=4,min_samples_split=5,random_state=42)
model.fit(X_train2, Y_train2)
acc = model.score(X_test2, Y_test2)
print(f"Test Accuracy: {acc}")

Test Accuracy: 0.7747440273037542


In [ ]:
dt = DecisionTreeClassifier(random_state=42)

param_grid_dt = {'max_depth': [3,5,7,8,10,None],'min_samples_split': [2,5,10],'min_samples_leaf': [1,2,3,5]}
grid_dt = GridSearchCV(dt,param_grid_dt,cv=5,scoring='accuracy',n_jobs=-1)
grid_dt.fit(X_train2, Y_train2)
print("Best DT Params:", grid_dt.best_params_)
print("Best DT Score:", grid_dt.best_score_)

Best DT Params: {'max_depth': 7, 'min_samples_leaf': 1, 'min_samples_split': 2}
Best DT Score: 0.7666666666666666


ok lastly after hyperparameter tuning & gridsearching

KNN: 76.58% ACCURACY &
DECISION TREE: 77.77%

lastly using random forest

In [ ]:
rf = RandomForestClassifier(n_estimators=100,max_depth=8,min_samples_split=5,min_samples_leaf=2,random_state=42)
rf.fit(X_train2, Y_train2)

Y_testpred_rf = rf.predict(X_test2)

print("Test Accuracy:", accuracy_score(Y_test2, Y_testpred_rf) * 100, "%")

#random forest showed 76.8% accuracy...lastly i will use grid search

Test Accuracy: 77.47440273037543 %


In [ ]:
param_grid_rf = {'n_estimators': [100, 200,250,280, 300,350,400],'max_depth': [5, 7,8,9, 10, None],'min_samples_split': [2,3,4, 5, 10],'min_samples_leaf': [1, 2,3]}

rf = RandomForestClassifier(random_state=42)

grid_rf = GridSearchCV(
    estimator=rf,param_grid=param_grid_rf,cv=5,scoring='accuracy',n_jobs=-1)

grid_rf.fit(X_train2, Y_train2)

print("Best RF Params:", grid_rf.best_params_)
print("Best CV Score:", grid_rf.best_score_)

Best RF Params: {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 4, 'n_estimators': 280}
Best CV Score: 0.7940170940170941
